### ISSUE TITLE
discussion about default value handling

### ISSUE THREAD

**Issue by avilleret** [ossia/libossia#614]
in ossia-max, when we instanciate an abstraction / load a patcher with some ossia.model or ossia.view, the default values are sent following the priority order : higher priority comes first.

the priority of the parent model/view is taken as follow (see here a patcher hierarchy, which is an ossia tree in this case)

```
    model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
    model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
```

The default values will be fired in the following order :

    /bar/c
    /bar/d
    /foo/a
    /foo/b

`/foo/a` comes after `/bar/d`even if 90 > -12 because their parent model's priority is taken into account.

Given that, when should we fire the value of object with absolute and global address ?

For instance : 

```
    model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
    model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
     |_ parameter /more/headache @priority 123
```

when should we fire the `/kk` and `/more/headache/ values ? 

I would say in that order : 

    /kk
    /bar/c
    /bar/d
    /more/headache
    /foo/a
    /foo/b

EDIT : /more/headache comes before /foo/a because more node's priority is 0 > -10

Because `kk`priority 98 > bar 20 > foo's -10 and more priority is not defined, so it's default to 0, then it show up between foo and bar.

Am I right ? What do people think ?
@jln- @matcham @bltzr @evanmtp @vincentgoudard

**Comment by avilleret** [comment]
to add some fun, what about a parameter with a global address ? (because yes, you can do that, at your own risk)

**Comment by jln-** [comment]
I guess I agree with what you describe. But to be honest, I had absolutely no idea one can declare an absolute address parameter under a model node :-$

**Comment by maybites** [comment]
I can't follow your logic. In my understanding:

```
/more/headache
/kk
/bar/c
/bar/d
/foo/a
/foo/b
```

if you want to prioritize according to the address, but I would prioritize according to the model-parameter relationship:

```
/more/headache
/bar/c
/bar/d
/kk
/foo/a
/foo/b
```

seems more logical to me. but then again prioritize according to the address gives more flexibility... hard call.

**Comment by avilleret** [comment]
@maybites : I really disagree with your first proposal
`/kk` has a lower priority than `/bar` so it should come after

Concerning the second, I understand your point and that's exactly why I need some feedback.
as it was specified, all nodes have a priority, which is 0. by default.
When we create a parameter from an absolute address, it will create all the intermediate nodes, giving them a default priority.
That's how it was specified.

If we take your second proposal, then when you have the following (each model in a separate subpatcher) :

    model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
    model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
     |_ parameter /more/headache @priority 123
    model more @priority 12
     |_ parameter test @priority 23

What would be the fire order @maybites ?

For me : 

    /kk
    /bar/c
    /bar/d
    /more/headache
    /more/test
    /foo/a
    /foo/b

Note that I made a mistake in my first post, I edited it.

**Comment by matcham** [comment]
Hey
I fully understand and agree with Antoine's first proposal.

after that... it's a bit confusing to me :-s 
No mater where the parameter is in a patch, only the resulting namespace counts, right ?
so I don't really get the point...

**Comment by vincentgoudard** [comment]
1) what are we talking about ?

@avilleret : can you explain what you mean by "absolute and global address" ? Are those two different things (absolute on one hand, global on the other) or a unique thing ?

> to add some fun, what about a parameter with a global address ? (because yes, you can do that, at your own risk)

isn't that what you refer to when writing "parameter /more/headache" in your first example ?

I just tried to create a "parameter /more/headache" in a submodel and it does appear at the root, so my understanding here is that this (non-documented?) syntax allow one to create parameters in subpatcher, that should be attached to the model in the patcher, to actually be at the root of the namespace, am I correct ?

2) (Optional question) : Does anyone have a concrete example in mind where such a technique is needed ?

3) If creating a "parameter /more/headache" (using this slashes syntax) brings the parameter to the root (IIUC), why not then having it displayed like this in your first example namespace (as I think it does in Max) :
```
 model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
    model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
parameter /more/headache @priority 123
```

=> In this case, the answer for priority seems obvious to me because /more/headache does not depend on any parent's priority.

4) What about param values that change the namespace like in the patcher below (At the moment, it crashes Max to recall presets in such a scheme) : 
```
<pre><code>
----------begin_max5_patcher----------
3387.3oc6cs0jahiE94t+UPQMyad5EciK6SyTyaSsOr09x9PlTtvFZ2jECt.
4jNSpr+1WgDXycD1HfNampR5zbw5b9z4hjNeR9aO9f9t3W8S0096ZeP6gG91
iO7.+RYW3g7e+A8itutOzMk+X56iOdzOhpuQbOp+qT90S8oZe1M7rep1ywIZ
6OmjvdrTsStItG8o9IoEuRz4iAQg9T9mGH+hAd7Ok3ce5W.Fkdx3yzhG0H+p
mbo6eIH5v1D+8TgnCLcdxXiFxBl8CfCH6GPimLz9X1678GeL6e1HoJxZ1c9I
xJtXn9EAKWU25G4tKzurPWUUf4WUbI5WO4KzCc8MZ56biNnq8wdzVCymHazv
HyL0jXvUV3rnrHmERYy6ZuSk84vXlrIsxZc6JKyG3nKWELuO02laXaLEZej+
WXpUCOWp1Ns.owDf935j4ZHSSCXwLjwnVno.iY0p1X1spssDtvFpqe8Th+I+
HOMWOuD+ToC+BsZuGFzoBKghRLLmjt010zzSILyqm0R+Zl.qs+E2nH+veNzq
3ICCh72GeNhVtmaPfvVA.Al3TBHPjYIPF.rBBjYx8uIllpyhONMMv8oD+iwT
+hatONLNQHBFOAfDryF1+wxwzvxN6+Y6fA.1+CjIJxZYPZ2x.0GBw+qLHDxJ
CarbtKD5Hyi28feCHJOXPgOBnUSHXKpLdRcFLIbsCjGU.oj3eZrHfoYxYa5n
QKtInoTGMsLx5NAXR1OLKYvW9c7SxUrbMi4vED5+Y1vkChiJ03On6d5ToK+P
oWICN9jvH2dykKEDItD7xkR7+bPw6ab4ptILImxD6yIb8R+Uyh95rOlXO+jn
yAbQQbQVGStHw6BhXATRO4tW7xY8TE2tT+Mx.K5o4iLGgyGNiyU3f0IeHLd+
+w2qr3oGyRgEDw6HintzZRutm+ytmCoaeNlMWif+hKCfrO7Vt+y4xXq2LSK3
JvukD3FdQENjD3EGkIDU5KxtbQywzNtwLfTVY3OQj6oVdYloFCX53loLk7b5
N2jrtp7.zE8f5z33vp25x6E5+LM+1mBXd1UQQZ7otuYRvgW54c2Eyt4w99r4
2Ic64Hwc2xrJnaSc+bUzl5FFl6oV8i+U2n.VJFeZfnK.Zb4lhjTujtOINLrh
9Jtyma4NdLq789eIvi9BugJaLvd7fSEFQ5W5k8BN3mRqdMp6gzpWIk9UAnW5
Rm2k6Euk5e7THSKp9.L+ifTZ5KweIM+AKLzJC.WmAdYu5xg8pb8dC0WMF3+8
uQYZF7oOkxz4ROPag6aLp3ZuPyHhCDUrqHiPjHdOKrOKxnXhcWC9+PQBf7XM
JCU.2DpfUFp.348AV.NpXuPnxsYq.TEp3Hxa3X9FzTApJPQXb.LwJzRo5570
XlkkFWUmCYtNdPL5FPLFv4fHVGj7ISVc48lAMlIIrzWku0ty6XIfxdfitIGB
DYwMjAQxdmJyJAsoz+TZdHMB83TV.5XHO8CsBoNMvyutcakAA8ObowZ+K+Cm
CcSz6umAXKV50RKjAwnlNHZ0RCgT48W7Qerv8V1qvdKB3ZmE1bkzY4F4M1PI
XqaNTBPrTThPIHGkDJQlzrzWoidrGJavGl7QjA.NKbZ1aATT13TAVbKEahBA
kJqcQ6qeqHaqVXraE2j1VSS48eLTLnAcr4nFp5xbtx8uT1vXgFyfCVO1R6ii
xVHWlj8hF6cz1EGSmNiodFnKrWXqVcgFxlhvmgjMTI1Tx6IVevE2E3gbTkMm
XQ1VZGwaJlNxTsfxRERWotgPaE6FlaPoTuPgvM1gCZdyiFzFKFFHeonDiKFU
eDwEics7J6wEfHO+W6ZfxWWJ3rN0NVfOt.kc+1Qlz3yI6K5kJF.nVUgyi4iE
DcYkw+vUaAMY6YFsTfjTJxRLnNo.JqTXnRo.rJvBCIkBSUJDqAjHKc5hiDHy
0fa5pvKsPCGTJJ53TiTrF5O.2uPjewhJvxWlKushBJt0kRSB1clJR1Ttjxip
vWGBi24FVqZUsUWrGuJbKJgk.vIkZBB94XO+7U5afmd5mC899cQZIaEfEnIg
xRiiRdVyNi73JoXwrUI6jtnOZ+tna+24cwJjsRX8wCW8BUVjqPEP7KpfSpoZ
6jl.mijOVEjRUJKCRdgYM4y3ACTmsQFYaB092Agd+SwJM7qkhbq8m5+ll2WY
QjC1qIdRWVH5+TWVPhLkgFH3bp1aUhPOSOCtN3S2dkfQsonnlJpyjpnhdenI
uPzPQdg+ekGW1BqeaNVXg3t9VlFuyhq2Yw06r3Rcr3Jg0vZFicQTU2ZnJFqD
Qroqv34dUTypDlHC9HWfzaDPpLBgg.EDOBIwd1qVnFKgS3189ggoilnRpqtW
VBPgiMlf54NaK+YU8uy7nckKs1q2VN0Nxq1ct0gyu1UN1J4YkHWas7shMKhM
G2LwBVaXWa4oaOkqDockI06foekLEbOogkKU7.oiGLk7fokGH07vomGLEsDo
okIU8XRW2SJ6ASa2ep69Se2eJ7dSi2Up71Sm2QJcoRqKYp81SuWOLS8H3Mt+
fo6qFNuBiZ6MPdez1t+34RDSuy35XNgby4REjypgpY6pjwqZVOUic6C8cSzp
xWyQAgNyADBrKuSWqujSJFBaLPh1WIVNaGpTN1dGrUWKhmZMEETa.h4C0vAs
FwQvshflyCBlyBRtAoIYMgf+UnVZHKMh1MGOjzODBGDB2HILl6BixiLtVcnu
USQ3rDTzJmBF7g7ZgWSXHUK7xJROZzCAl.yvFSDsS57ZK15olWY68ORVh1yg
onfRzqRKwBLz3lQv4IybdLv0XdkIZDNPyYLp3pbHN7JT6mzK.R5A.w8Cf3g.
vLhKt45IrPcdL18tsTrFChPjlfYER4HSuHlQ2HFXV7cEYPvFWSfTmVj8QMxN
nG4pERu2DJRmXNeReSCtBmAbsAebGW3Qi9AVCIPLD7ID6OXrg3O..CyrQp1v
r5JO1fCuMw01nCVKnaCZg0fvh8QMrZTYcTc72p.ZMFAbAjOyQHeVKf7QFC9Y
uPBHXE2Ai.i.AIKk.JMBhWBIbL1fYBHZlEvl7DdsEFDNp3flKgDZt1iz.wiP
BaWahS7D0J1X8J4cXpbQzAyqnu18r.iw22YMAsiRxQqI6YvXhlUj.t2gK+Xa
7cXncnPcxdbSUEt+crPa6ZgpJfp2ymggZWH7gFXrrVBzCOtvCssNabvS1OQc
vEGHivVYpyH6L8OtqEpKbh0nWnXioZ5NFjLc+DXzmvUJ6r2H+.zQbROSLmah
iQ07Ne7jVnezA5Ki137V25+YsYlgYdyJm4IznDenl28s8c5CClCeXggDTriL
.FVus8gGjqiZY8Ji1M1PcmeevRvO4ctN9NWGemqiuy0w2453v7p3lqiM.emE
tQ5xEZI13SWJ85LVsv6qDL.z8WBFGhnNKNuUpF3PLDMqp++zMynLvrPEJnfO
Y.Gi111ouwIfBYVnMgvtERD7OA81oF+2aXMNCnGSfMn8akBRqV5nfVH1nj6q
icV.1nnRZOBmSVOl6ummjdd2SB2G+Slyng1+vvbhwTXc.ZAJrwXpYDXsWVXv
RT5e3Jux+vQT3+kvBDfGYApfys.t16gASJ6mduLYpZI1wyRUxDqwKx7GixjU
ePKcNfko3qEHoNEED04APL6dPJ21ww5rBQ.EWiAfo88fOv4pvhhMh6tQeNcP
twhJJl16FIs0PD9Y+KDfqWvl4AeRXxuuV8S2J4qjE4NNtmktZVv1+FZbQC.g
UYDnbqh6zECs3GI1Hxs+ksjP2MwfoNJrRNTrk8v1EQT4gsqzmL2jUvQ9KPoG
4uDYO3ga7bMow2zKY.Y5jZPna8iAdmhChn4FsHGdfQr37OzRraox9kkFTyH3
PMQuIECmVQCrFb.gqfy8afzm62.UJEixTQYRAdMXU.j0r.XpZoPl3N.kZWHq
eZ8khbZkBiwkcXQOJ1AjezOJ1656I99Ozca4KHdfBNzcAhkCv5995g+SAzmx
NVD2kwbrBorE9coWh5OFxp0lk+.aRpr1WIJNGjKiN6iC6+L4NLHk15JUo2IO
z5AWQBRmgf4mow1kNMiYhxEZLAwFWupm+yhOjOvDaMSyMY+fO2fq+CeBCcb4
eAT7WsOpxiU87SSeUdhpCm5STcwYGZ92rR.b6G+77OfpSbR.a0i9kCdCrME6
JcboAEznCR1Vp52MORjr6laop6MTI9FN4laIGYToJC.r9bFHP9QCEzguzJk+
s6TzrkQzlBLvRhFppvTcFcJsgws2vF2UCSjngcDtU2mGiDsi0T3uHiYLZR7L
koOaJLJQFxzP1SQKIowvTzNPIZG3DzNCkOfLA5CTl.Tvov7FJkQ2jnSj4JEG
DOSNR.YrumBrC.mKMBLByg1SZm+cOKwAKpLILKocseSbluW7jHvcmPWFwF0.
eDCPr19DJqIpsufpsefZtOf5d++Tee+vm9qXSITaPoWlSxe3dL9n61WBNvz7
CuPylkbhaTwbNK1PMUQD2868inkF6NPLYBx0YTTn5axQ1D1KTe2Qnej8IGTL
JfBDpiskzie+w+G.1xLuH
-----------end_max5_patcher-----------
</code></pre>
```

**Comment by maybites** [comment]
> 2. (Optional question) : Does anyone have a concrete example in mind where such a technique is needed ?

not in my case

> If creating a "parameter /more/headache" (using this slashes syntax) brings the parameter to the root (IIUC), why not then having it displayed like this in your first example namespace (as I think it does in Max) :

There is now some confusion:

I think the diagrams indentation are interpreted in different ways:

@avilleret I assume you mean this

```
model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
     |_ parameter /more/headache @priority 123
```

and not this (as in @vincentgoudard post indicated)

```
model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
     model bar @priority 20
          |_ parameter c @priority 30
          |_ parameter d @priority -12
          |_ parameter /more/headache @priority 123
```

in case of 

```
model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
     |_ parameter /more/headache @priority 123
model more @priority 12
     |_ parameter test @priority 23
```

and assuming 
1. higher priority = earlier fire
2. absolute address giving them a default priority

I concur with @vincentgoudard  that the diagram could be written like this:

```
model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
parameter /kk @priority 98
model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
parameter /more/headache @priority 123
model more @priority 12
     |_ parameter test @priority 23
```

where the order would be

```
/more/headache
/kk
/bar/c
/bar/d
/more/test
/foo/a
/foo/b
```

so @avilleret I stick to my order from my first post. But it could very well be I misunderstand something fundamental.

**Comment by avilleret** [comment]
@vincentgoudard 

1. `foo/bar` is a relative address, `/foo/bar` is an absolute one (because it starts with `/`), `dev_name:/foo/bar` is a global one (with device name prefix)

I removed the comment, which is irrelevant (posted before editing first post...)

> I just tried to create a "parameter /more/headache" in a submodel and it does appear at the root, so my understanding here is that this (non-documented?) syntax allow one to create parameters in subpatcher, that should be attached to the model in the patcher, to actually be at the root of the namespace, am I correct ?

you're totally right !

2. you might want to add a parameter to a model without modifying you model patcher, or you might want to add a parameter to a specific instance of a model, in that case absolute address might be useful

3. my diagram was meant as Max patcher hierarchy, each model in a subpatcher and parameter `/more/headache` in one of those subpatchers. But it could be anywhere (under the same device) since it has an absolute address.
The namespace is (sorted by priority): 

    /kk
    /bar/c
    /bar/d
    /more/headache
    /more/test
    /foo/a
    /foo/b


4. @vincentgoudard  could open another issue with your crashing patcher ? thanks

**Comment by avilleret** [comment]
from @maybites 

> absolute address giving them a default priority

no, absolute address creates intermediate node, i.e,  `[parameter /more/headache]`, would create the node `more` then the node `headache` and create a parameter on the later.
Each created node have a priority, which is default to 0.
That's why, the node `more` have a priority of 0 and will be fired after `bar` and before `foo`.
Priorities are only compare to the same level of hierarchy.

**Comment by vincentgoudard** [comment]
@avilleret : thanks for the clarification!
Then I agree with your proposal on priority.

Regarding my 4th point, I'll create an issue for that one, will try to sort out first what bit makes it crash. But regardless of the crash, the question remains : how would ossia deal with preset recall in this case ?

**Comment by maybites** [comment]
.. ok, I get it. I think this is too confusing.

I propose a simple rule:

`priorities follow the model/parameter hierarchies: first the priorities of the models are sorted, then the priorities of the properties.  
`

your current proposal will confuse people and let them think it is a bug. Unless you make a really really comprehensive help patch that explicitly explains the in and outs of the priorities. 

Personally I am more inclined for simple rules and not complicated documentation. Unless important functionality is compromised.

**Comment by vincentgoudard** [comment]
> I propose a simple rule:
> 
> ``priorities follow the model/parameter hierarchies: first the priorities of the models are sorted, then the priorities of the properties. ``

@maybites isn't it precisely what @avilleret was proposing in his first post ?

This is also what I think : let the priorities be sorted as they are represented in the namespace, regardless where the actual `ossia` object is located in the patch.

**Comment by maybites** [comment]
> isn't it precisely what @avilleret was proposing in his first post ?

to my understanding he asked 

`... when should we fire the value of object with absolute and global address ?`

and proposed then the solution that kept me confused for a moment.

what I don't like about the current proposal is in case of

```
model foo @priority -10
     |_ parameter a @priority 100
     |_ parameter b @priority 90
     |_ parameter /kk @priority 98
model bar @priority 20
     |_ parameter c @priority 30
     |_ parameter d @priority -12
     |_ parameter /more/headache @priority 123
model more @priority 12
     |_ parameter test @priority 23
```

the fact that 

`model more @priority 12`

has an influence on the parameter '/more/headache' which is located under another model.

```
model bar @priority 20
     |_ parameter /more/headache @priority 123
```

this is not changing the priority of 'more', only the priority of 'headache' - thats what threw me off the rails.

this is not intuitive. 

so in conclusion

@avilleret proposes a priority (default = 0) for each node, and then simply sorting through the node tree. or as @vincentgoudard stated it 'let the priorities be sorted as they are represented in the namespace, regardless where the actual ossia object is located in the patch.'

I propose a priority (default = 0) for each [model] or [parameter] object and then sort through the patcher based model - parameter hierarchy.

But I understand @avilleret 's implementation - I guess its easier to realize - once the nodetree is created its simply sorting through the node tree. While my dumbyfing it will make this task a mess.

**Comment by vincentgoudard** [comment]
@maybites  : OK I understand your point. 
To me though,  `parameter /more/headache @priority 123` is not attached to model `bar`. 
It just happens to be in the same patcher.

> what I don't like about the current proposal is (...) the fact that
> > model more @priority 12
> has an influence on the parameter '/more/headache' which is located under another model.

Actually, the reverse would be even worse for me, if `parameter /more/headache @priority 123` was modifying the priority for node `/more` and/or `/headache`... all their potential subnodes!

Regarding the situation where this would happen, I understand the example you gave @avilleret but they only describe the reverse situation, that is, creating a parameter at root patcher for example, that is added (maybe temporarily) to a model, thanks to providing the correct path to it. But then, how do you receive it in the abstraction ?

Ossia parameters let you (potentially) do crazy things like creating a single `ossia.parameter` and using wildcard syntax like `address {frequency|resonance|pitch|duration|gain|filter_coef|whatnot}` to have all your parameters arrive at a single location in your patch. \[Actually it fails doing so, see #616 . \] In this regard, priorities relate to parameter scheduling and to a large extent to initialization questions and this is where I can see that `ossia.parameter` objects location do matter.

Sorry if this comment seems to hijack the thread a bit, but I think it's always good not to discuss actual usecases when dealing with such complex topics.

**Comment by vincentgoudard** [comment]
OK, let me simplify my own thoughts and say that I agree with @avilleret proposal **as long as** there is a notification at the end of the values recall (e.g. `done`).
😄

**Comment by avilleret** [comment]
the value recall is done in one scheduler tick, so using a well-known `[trigger done bang]` should do the trick :-)

> how would ossia deal with preset recall in this case ?

concerning the preset recall order, if we speak about ossia's preset system, then, the value order follow the order in the preset file. And this order follow priorities on saving. 
This was a request at some point: to be able to control preset value order by editing the preset.

**Comment by avilleret** [comment]
this have been implemented, so I'm closing the issue
